src/uimports.cpp

```diff
@@ -106,6 +106,34 @@ static BOOL UPDATE_IMPORTS_XX(HANDLE hProcess,
         }
     }
 
+	///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+	//Added code
+	if (inh.IMPORT_DIRECTORY.VirtualAddress!=NULL && inh.IMPORT_DIRECTORY.Size==0)
+	{
+		//Don't worry about changing the PE file, because the load information of the original PE header has been saved and will be restored. The change here is just for the following code to work normally
+		PIMAGE_IMPORT_DESCRIPTOR pImageImport=(PIMAGE_IMPORT_DESCRIPTOR)(pbModule+inh.IMPORT_DIRECTORY.VirtualAddress);
+		IMAGE_IMPORT_DESCRIPTOR ImageImport;
+
+		if(!ReadProcessMemory(hProcess,pImageImport,&ImageImport,sizeof(ImageImport),NULL))
+		{
+			_Analysis_assume_(FALSE);
+			goto finish;
+		}
+		while(ImageImport.Name)
+		{
+			inh.IMPORT_DIRECTORY.Size+=sizeof(IMAGE_IMPORT_DESCRIPTOR);
+			if(!ReadProcessMemory(hProcess,pImageImport,&ImageImport,sizeof(ImageImport),NULL))
+			{
+				_Analysis_assume_(FALSE);
+				goto finish;
+			}
+			++pImageImport;
+		}
+		OutputDebugString(TEXT("[This PE file has an import table, but the import table size is marked as 0. This is an error.")
```

Comment:
We should use DETOUR_TRACE instead of OutputDebugStrring?

Reply:
DETOUR_TRACE can not ouput any message if no defined DETOUR_DEBUG, I use OutputDebugString to tell user to pay attention to this is a unnormal exe.

Reply:
Should finish with a ] if it starts with [. Also OutputDebugString doesn't automatically append a new line, so you want to add `\n` at the end of the string.

Reply:
`\r\n` is in next line, this is a const string concat operation.

Reply:
> DETOUR_TRACE can not output any message if no define DETOUR_DEBUG, I use OutputDebugString to force output message for tell user to pay attention to this is a unnormal exe.

So the users is going to just happen to be running under a debugger? 
OutputDebugString only outputs a message if a debugger is attached.

It's not really detours job to emit messages as library.
I think it would make more sense to just return a rich
error code in LastError, and let the caller handle that case
explicitly.

Reply:
[Dbgview ](https://docs.microsoft.com/en-us/sysinternals/downloads/debugview) and many debug tools can capture the debug output statements of the application without being debugged.
In fact, it doesn't matter whether the user notices this message, it's like an MFC running warning.
like 
```c++
		TRACE(traceAppMsg, 0, _T("Warning: calling DestroyWindow in CWnd::~CWnd; ")
		   _T("OnDestroy or PostNcDestroy in derived class will not be called.\n"));
```
in
```c++
CWnd::~CWnd()
{
	if (m_hWnd != NULL &&
		this != (CWnd*)&wndTop && this != (CWnd*)&wndBottom &&
		this != (CWnd*)&wndTopMost && this != (CWnd*)&wndNoTopMost)
	{
		TRACE(traceAppMsg, 0, _T("Warning: calling DestroyWindow in CWnd::~CWnd; ")
		   _T("OnDestroy or PostNcDestroy in derived class will not be called.\n"));
		DestroyWindow();
	}
```
The user cannot handle it because the only way to deal with it is to modify the PE file on the disk, but this file may contain a digital signature, and the modification may be illegal, and it only has a problem when it is launched through Detours' createwith API. The user uses Explorer. exe starts it, or directly calls the CreateProcess function to start it, there is no problem, these are enough to prove a defect of Detours.
The only improvement is that calling OutputDebugString in a non-debugging state may modify the value of LastError of the current thread, which has been fixed in the latest submission.